rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashSet};
+
+use arrow_array::RecordBatch;
+use arrow_schema::Schema as ArrowSchema;
+use object_store::path::Path;
+use path_absolutize::Absolutize;
+use tempfile::TempDir;
+
+use crate::datatypes::Schema;
+use crate::io::reader::batches_stream;
+use crate::io::FileReader;
+use crate::{
+    error::Result,
+    io::{FileWriter, ObjectStore, RecordBatchStream},
+};
+
+const BUFFER_FILE_NAME: &str = "buffer.lance";
+
+/// Shuffle [RecordBatch] based on their IVF partition.
+///
+///
```

Comment:
Could we add a bit more commentary here. I was thinking at first we wrote a file per partition, and then later realized it's all one file with the groups.

```suggestion
/// This buffers all the batches into a single Lance file, but keeps
/// track of which Lance file groups belong to each partition.
```

Reply:
Done